File: src/contests/cg/puzzle/fsharp/telephone-numbers.fs
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

let n = int (System.Console.ReadLine().Trim())
let root = System.Collections.Generic.Dictionary<char, obj>()
let mutable cable = 0
for _ in 1 .. n do
    let number = System.Console.ReadLine().Trim()
    let mutable node = root
    for d in number do
        if not (node.ContainsKey d) then
            node.[d] <- System.Collections.Generic.Dictionary<char, obj>() :> obj
            cable <- cable + 1
        node <- node.[d] :?> System.Collections.Generic.Dictionary<char, obj>
printfn "%d" cable

File: src/contests/cg/puzzle/csharp/telephone-numbers.cs
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

using System;
using System.Collections.Generic;

class Solution
{
    static void Main()
    {
        int n = int.Parse(Console.ReadLine().Trim());
        var root = new Dictionary<char, object>();
        int cable = 0;
        for (int i = 0; i < n; i++)
        {
            string number = Console.ReadLine().Trim();
            var node = root;
            foreach (char d in number)
            {
                if (!node.ContainsKey(d))
                {
                    node[d] = new Dictionary<char, object>();
                    cable++;
                }
                node = (Dictionary<char, object>)node[d];
            }
        }
        Console.WriteLine(cable);
    }
}
